nes-network/bindings/network/lib.rs

```diff
@@ -0,0 +1,359 @@
+        #[allow(non_snake_case)]
+        fn identifyThread(thread_name: &str, worker_id: &str);
+    }
+
+    extern "Rust" {
+        type ReceiverNetworkService;
+        type SenderNetworkService;
+        type SenderDataChannel;
+        type ReceiverDataChannel;
+
+        fn enable_memcom() -> Result<()>;
+        fn init_receiver_service(connection_addr: String, worker_id: String) -> Result<()>;
+        fn receiver_instance(connection_addr: String) -> Result<Box<ReceiverNetworkService>>;
+        fn init_sender_service(connection_addr: String, worker_id: String) -> Result<()>;
+        fn sender_instance(connection_addr: String) -> Result<Box<SenderNetworkService>>;
+
+        fn register_receiver_channel(
+            server: &mut ReceiverNetworkService,
+            channel_identifier: String,
+        ) -> Box<ReceiverDataChannel>;
+        fn receive_buffer(
+            receiver_channel: &ReceiverDataChannel,
+            builder: Pin<&mut TupleBufferBuilder>,
+        ) -> Result<bool>;
+        fn interrupt_receive(receiver_channel: &ReceiverDataChannel);
+
+        fn close_receiver_channel(channel: Box<ReceiverDataChannel>);
+
+        fn register_sender_channel(
+            server: &SenderNetworkService,
+            connection_identifier: String,
+            channel_identifier: String,
+        ) -> Result<Box<SenderDataChannel>>;
+
+        fn close_sender_channel(channel: Box<SenderDataChannel>);
+        fn flush_channel(channel: &SenderDataChannel) -> bool;
+        fn try_send_on_channel(
+            channel: &SenderDataChannel,
+            metadata: SerializedTupleBuffer,
+            data: &[u8],
+            children: &[&[u8]],
+        ) -> SendResult;
+    }
+}
+
+#[derive(Clone)]
+enum ReceiverService {
+    MemCom(Arc<receiver::NetworkService<channel::MemCom>>),
+    Tcp(Arc<receiver::NetworkService<channel::TcpCommunication>>),
+}
+#[derive(Clone)]
+enum SenderService {
+    MemCom(Arc<sender::NetworkService<channel::MemCom>>),
+    Tcp(Arc<sender::NetworkService<channel::TcpCommunication>>),
+}
+#[derive(Default)]
+struct Services {
+    receiver: HashMap<ThisConnectionIdentifier, ReceiverService>,
+    sender: HashMap<ThisConnectionIdentifier, SenderService>,
+}
+
+static USE_MEMCOM: Mutex<bool> = Mutex::new(false);
+fn enable_memcom() -> Result<(), String> {
+    let mut use_memcom = USE_MEMCOM.lock().unwrap();
+    if *use_memcom {
+        warn!("Memcom is already enabled");
+        // return Err("Memcom is already enabled".into());
+    }
+    *use_memcom = true;
+    Ok(())
+}
+/// Lazy singleton types, initialized on first use in a thread-safe way
+static SERVICES: sync::Lazy<Mutex<Services>> = sync::Lazy::new(|| Mutex::default());
+
+pub struct ReceiverNetworkService {
+    handle: ReceiverService,
+}
+
+struct SenderNetworkService {
+    handle: SenderService,
+}
+
+struct SenderDataChannel {
+    chan: SenderChannel,
+}
+
+struct ReceiverDataChannel {
+    chan: Box<ReceiverChannel>,
```

Comment:
Why is this one Boxed and the sender channel not?

Reply:
that is a good question, the authors of async_channel decided to make the receiver explicitly not Unpin, which means it cannot be moved and thus needs to be boxed